### ISSUE TITLE
Rule Change: `no-unused-vars` should consider this is used

### ISSUE THREAD

**Issue by fisker** [eslint/eslint#17299]
### What rule do you want to change?

no-unused-vars

### What change to do you want to make?

Generate fewer warnings

### How do you think the change should be implemented?

Other

### Example code

```js
let promise;

function doSomeThingOnce() {
  promise ??= doSomeThing();
}

function doSomeThing() {}
doSomeThingOnce();
doSomeThingOnce();
```

### What does the rule currently do for this code?

```
Error 1:5 'promise' is assigned a value but never used.  ([no-unused-vars](https://eslint.org/docs/rules/no-unused-vars))
```

[playground](https://eslint.org/play/#eyJ0ZXh0IjoibGV0IHByb21pc2U7XG5cbmZ1bmN0aW9uIGRvU29tZVRoaW5nT25jZSgpIHtcbiAgcHJvbWlzZSA/Pz0gZG9Tb21lVGhpbmcoKTtcbn1cblxuZnVuY3Rpb24gZG9Tb21lVGhpbmcoKSB7fVxuZG9Tb21lVGhpbmdPbmNlKCk7XG5kb1NvbWVUaGluZ09uY2UoKTsiLCJvcHRpb25zIjp7InBhcnNlck9wdGlvbnMiOnsiZWNtYVZlcnNpb24iOiJsYXRlc3QiLCJzb3VyY2VUeXBlIjoic2NyaXB0IiwiZWNtYUZlYXR1cmVzIjp7fX0sInJ1bGVzIjp7ImNvbnN0cnVjdG9yLXN1cGVyIjpbImVycm9yIl0sImZvci1kaXJlY3Rpb24iOlsiZXJyb3IiXSwiZ2V0dGVyLXJldHVybiI6WyJlcnJvciJdLCJuby1hc3luYy1wcm9taXNlLWV4ZWN1dG9yIjpbImVycm9yIl0sIm5vLWNhc2UtZGVjbGFyYXRpb25zIjpbImVycm9yIl0sIm5vLWNsYXNzLWFzc2lnbiI6WyJlcnJvciJdLCJuby1jb21wYXJlLW5lZy16ZXJvIjpbImVycm9yIl0sIm5vLWNvbmQtYXNzaWduIjpbImVycm9yIl0sIm5vLWNvbnN0LWFzc2lnbiI6WyJlcnJvciJdLCJuby1jb25zdGFudC1jb25kaXRpb24iOlsiZXJyb3IiXSwibm8tY29udHJvbC1yZWdleCI6WyJlcnJvciJdLCJuby1kZWJ1Z2dlciI6WyJlcnJvciJdLCJuby1kZWxldGUtdmFyIjpbImVycm9yIl0sIm5vLWR1cGUtYXJncyI6WyJlcnJvciJdLCJuby1kdXBlLWNsYXNzLW1lbWJlcnMiOlsiZXJyb3IiXSwibm8tZHVwZS1lbHNlLWlmIjpbImVycm9yIl0sIm5vLWR1cGUta2V5cyI6WyJlcnJvciJdLCJuby1kdXBsaWNhdGUtY2FzZSI6WyJlcnJvciJdLCJuby1lbXB0eSI6WyJlcnJvciJdLCJuby1lbXB0eS1jaGFyYWN0ZXItY2xhc3MiOlsiZXJyb3IiXSwibm8tZW1wdHktcGF0dGVybiI6WyJlcnJvciJdLCJuby1leC1hc3NpZ24iOlsiZXJyb3IiXSwibm8tZXh0cmEtYm9vbGVhbi1jYXN0IjpbImVycm9yIl0sIm5vLWV4dHJhLXNlbWkiOlsiZXJyb3IiXSwibm8tZmFsbHRocm91Z2giOlsiZXJyb3IiXSwibm8tZnVuYy1hc3NpZ24iOlsiZXJyb3IiXSwibm8tZ2xvYmFsLWFzc2lnbiI6WyJlcnJvciJdLCJuby1pbXBvcnQtYXNzaWduIjpbImVycm9yIl0sIm5vLWlubmVyLWRlY2xhcmF0aW9ucyI6WyJlcnJvciJdLCJuby1pbnZhbGlkLXJlZ2V4cCI6WyJlcnJvciJdLCJuby1pcnJlZ3VsYXItd2hpdGVzcGFjZSI6WyJlcnJvciJdLCJuby1sb3NzLW9mLXByZWNpc2lvbiI6WyJlcnJvciJdLCJuby1taXNsZWFkaW5nLWNoYXJhY3Rlci1jbGFzcyI6WyJlcnJvciJdLCJuby1taXhlZC1zcGFjZXMtYW5kLXRhYnMiOlsiZXJyb3IiXSwibm8tbmV3LXN5bWJvbCI6WyJlcnJvciJdLCJuby1ub25vY3RhbC1kZWNpbWFsLWVzY2FwZSI6WyJlcnJvciJdLCJuby1vYmotY2FsbHMiOlsiZXJyb3IiXSwibm8tb2N0YWwiOlsiZXJyb3IiXSwibm8tcHJvdG90eXBlLWJ1aWx0aW5zIjpbImVycm9yIl0sIm5vLXJlZGVjbGFyZSI6WyJlcnJvciJdLCJuby1yZWdleC1zcGFjZXMiOlsiZXJyb3IiXSwibm8tc2VsZi1hc3NpZ24iOlsiZXJyb3IiXSwibm8tc2V0dGVyLXJldHVybiI6WyJlcnJvciJdLCJuby1zaGFkb3ctcmVzdHJpY3RlZC1uYW1lcyI6WyJlcnJvciJdLCJuby1zcGFyc2UtYXJyYXlzIjpbImVycm9yIl0sIm5vLXRoaXMtYmVmb3JlLXN1cGVyIjpbImVycm9yIl0sIm5vLXVuZGVmIjpbImVycm9yIl0sIm5vLXVuZXhwZWN0ZWQtbXVsdGlsaW5lIjpbImVycm9yIl0sIm5vLXVucmVhY2hhYmxlIjpbImVycm9yIl0sIm5vLXVuc2FmZS1maW5hbGx5IjpbImVycm9yIl0sIm5vLXVuc2FmZS1uZWdhdGlvbiI6WyJlcnJvciJdLCJuby11bnNhZmUtb3B0aW9uYWwtY2hhaW5pbmciOlsiZXJyb3IiXSwibm8tdW51c2VkLWxhYmVscyI6WyJlcnJvciJdLCJuby11bnVzZWQtdmFycyI6WyJlcnJvciJdLCJuby11c2VsZXNzLWJhY2tyZWZlcmVuY2UiOlsiZXJyb3IiXSwibm8tdXNlbGVzcy1jYXRjaCI6WyJlcnJvciJdLCJuby11c2VsZXNzLWVzY2FwZSI6WyJlcnJvciJdLCJuby13aXRoIjpbImVycm9yIl0sInJlcXVpcmUteWllbGQiOlsiZXJyb3IiXSwidXNlLWlzbmFuIjpbImVycm9yIl0sInZhbGlkLXR5cGVvZiI6WyJlcnJvciJdfSwiZW52Ijp7ImVzNiI6dHJ1ZX19fQ==)

### What will the rule do after it's changed?

Should not report this case.

### Participation

- [ ] I am willing to submit a pull request to implement this change.

**Comment by Rec0iL99** [comment]
Hi @fisker, can you please provide more context as to why this change is needed? 

The variable `promise` is still unused here in the example and ESLint is right in reporting this right? Am I missing something here?

Btw, you can bypass this by using the `{ vars: "local" }}` option.

[playground](https://eslint.org/play/#eyJ0ZXh0IjoiLyogZXNsaW50IG5vLXVudXNlZC12YXJzOiBbXCJlcnJvclwiLCB7XCJ2YXJzXCI6IFwibG9jYWxcIn1dICovXG5cbmxldCBwcm9taXNlO1xuXG5mdW5jdGlvbiBkb1NvbWVUaGluZ09uY2UoKSB7XG4gIHByb21pc2UgPz89IGRvU29tZVRoaW5nKCk7XG59XG5cbmZ1bmN0aW9uIGRvU29tZVRoaW5nKCkge31cbmRvU29tZVRoaW5nT25jZSgpO1xuZG9Tb21lVGhpbmdPbmNlKCk7Iiwib3B0aW9ucyI6eyJwYXJzZXJPcHRpb25zIjp7ImVjbWFWZXJzaW9uIjoibGF0ZXN0Iiwic291cmNlVHlwZSI6InNjcmlwdCIsImVjbWFGZWF0dXJlcyI6e319LCJydWxlcyI6e30sImVudiI6eyJlczYiOnRydWV9fX0=)

**Comment by aladdin-add** [comment]
technically it's a bug to me.

```js
promise ??= doSomeThing();

// same as
promise = promise ?? doSomeThing();
```

`promise` was assigned a value and its value was actually read and used.

**Comment by fisker** [comment]
@Rec0iL99  Let me change the question. How should I fix my case?

**Comment by Rec0iL99** [comment]
@aladdin-add wouldn't then this case also be a bug?

```js
let something;

something = something + 1;
something = something + 1;
```

> @Rec0iL99 Let me change the question. How should I fix my case?

You would have to use `promise` by passing it to an API call, a function, etc. right? Technically, you are just assigning a value to `promise` and that is what the error message is pointing out: `'promise' is assigned a value but never used`

**Comment by Rec0iL99** [comment]
> ```js
> // same as
> promise = promise ?? doSomeThing();
> promise = promise ?? doSomeThing();
> ```

Ah, ok I get your point now. My apologies.

**Comment by gweesin** [comment]
@Rec0iL99 It looks interesting. The `promise` variable is not actually used based on the result, but from the perspective of reading and writing the variable, it is indeed being used.

So, could I consider this to be a valid expression?

Now I intend to make these two codeblocks consistent, but I have encountered some difficulties. I have identified the problem in the `isUnusedExpression` method, but I am currently unsure how to handle this scenario and make it compatible with all shorthand assignment statements.

```js
/**
 * Checks whether a given node is unused expression or not.
 * @param {ASTNode} node The node itself
 * @returns {boolean} The node is an unused expression.
 * @private
 */
function isUnusedExpression(node) {
    const parent = node.parent;

    if (parent.type === "ExpressionStatement") {
        // TODO we need to update this method or take some action here to resolve this question
        return true;
    }

    if (parent.type === "SequenceExpression") {
        const isLastExpression = parent.expressions[parent.expressions.length - 1] === node;

        if (!isLastExpression) {
            return true;
        }
        return isUnusedExpression(parent);
    }

    return false;
}
```

**Comment by gweesin** [comment]
😕 I apologize for my impatience. We should first discuss which segment's result should be considered normal before considering how to modify it.

**Comment by aladdin-add** [comment]
IMHO, it should be seen as used for `promise = promise ?? doSomething()` and `promise ??= doSomething()`, so no linting errors should be reported. marking as accepted.

**Comment by mdjermanovic** [comment]
I think `promise` should be considered used if there's a `promise ??=`, `promise ||=`, or `promise &&=` reference.

```js
var promise1, promise2;

promise1 ?? x; // this is considered using `promise1`
promise2 ??= x; // so this should be considered using `promise2`
```

The fix would check the assignment operator here:

https://github.com/eslint/eslint/blob/9718a9781d69d2c40b68c631aed97700b32c0082/lib/rules/no-unused-vars.js#L466-L470

@gweesin would you like to work on this?

**Comment by gweesin** [comment]
@mdjermanovic Yes, I'm glad to have the opportunity to try and resolve this issue.

**Comment by Skalman** [comment]
> If we expect `promise2 ??= x;` to be considered used, should `promise2 += x;` to be considered used?

No. In `promise2 ??= x`, `x` is conditionally evaluated, but in `promise2 += x`, `x` is _always_ evaluated, making `promise2` unnecessary.

```js
// These are equivalent
promise2 ??= x;

if (promise2 == null) { // <-- it's used here
  promise2 = x;
}
```

Such an `if` statement is hidden in `??=`, `||=` and `&&=`, but none of the other assignment operators.

**Comment by gweesin** [comment]
Thank you for the explanation. I think I understand why we need to do it this way. I will consider how to adjust the code accordingly.
